parser/src/parser/grammar.lalrpop

```diff
@@ -57,399 +111,318 @@ Trace: Vec<Vec<TraceBinding>> = {
+    <l:@L> "boundary_constraints" ":" <body: StatementBlock> <r:@R> => Span::new(span!(l, r), body),
 }
 
 Boundary: Boundary = {
     "first" => Boundary::First,
     "last" => Boundary::Last
 }
 
-// --- BOUNDARY CONSTRAINT EXPRESSIONS WITH PRECEDENCE (LOWEST TO HIGHEST) ----------------------
-
-BoundaryExpr: Expression = {
-    <lexpr: BoundaryExpr> "+" <rexpr: BoundaryFactor> =>
-        Expression::Add(Box::new(lexpr), Box::new(rexpr)),
-    <lexpr: BoundaryExpr> "-" <rexpr: BoundaryFactor> =>
-        Expression::Sub(Box::new(lexpr), Box::new(rexpr)),
-    BoundaryFactor
-}
-
-BoundaryFactor: Expression = {
-    <lexpr: BoundaryFactor> "*" <rexpr: BoundaryExponent> =>
-        Expression::Mul(Box::new(lexpr), Box::new(rexpr)),
-    BoundaryExponent
-}
-
-BoundaryExponent: Expression = {
-    <lexpr: BoundaryExponent> "^" <rexpr: BoundaryAtom> =>
-        Expression::Exp(Box::new(lexpr), Box::new(rexpr)),
-    BoundaryAtom
-}
+// INTEGRITY CONSTRAINTS
+// ================================================================================================
 
-BoundaryAtom: Expression = {
-    "(" <BoundaryExpr> ")",
-    <n: Num_u64> => Expression::Const(n),
-    <symbol_access: SymbolAccess> => Expression::SymbolAccess(symbol_access),
-    <list_folding_type: ListFolding<BoundaryExpr>> => Expression::ListFolding(list_folding_type),
+IntegrityConstraints: Span<Vec<Statement>> = {
+    <l:@L> "integrity_constraints" ":" <body: StatementBlock> <r:@R>
+        => Span::new(span!(l, r), body)
 }
 
-// INTEGRITY CONSTRAINTS
+// CONSTRAINTS
 // ================================================================================================
 
-IntegrityConstraints: Vec<IntegrityStmt> = {
-    "integrity_constraints" ":" <integrity_stmts: IntegrityStmts> => integrity_stmts
+Statement: Statement = {
+    <Let> => Statement::Let(<>),
+    ConstraintStatement,
 }
 
-IntegrityStmts: Vec<IntegrityStmt> = {
-    <l:@L> <integrity_stmts_groups: IntegrityStmtGroup+> <r:@R> =>? {
-        let integrity_stmts: Vec<IntegrityStmt> = integrity_stmts_groups.into_iter().flatten().collect();
-        // check if at least one integrity constraint is defined
-        let integrity_constraints_exist = integrity_stmts.iter().any(|stmt| match stmt {
-            IntegrityStmt::Constraint(_) => true,
-            _ => false,
-        });
-        if !integrity_constraints_exist {
-            diagnostics.diagnostic(Severity::Error)
-                .with_message("at least one integrity constraint must be declared")
-                .with_primary_label(span!(l, r), "missing constraint declaration in this section")
-                .emit();
-            return Err(ParseError::Failed.into());
-        }
-        Ok(integrity_stmts)
+StatementBlock: Vec<Statement> = {
+    <Let> => vec![Statement::Let(<>)],
+    <stmts:ConstraintStatement+> <last:Let> => {
+        let mut stmts = stmts;
+        stmts.push(Statement::Let(last));
+        stmts
     },
```

Comment:
For sure, always happy to answer questions to make review easier if it takes some of the load off :)

This case handles when a statement block contains a `let`, but starts with a non-let statement. `let` is always the last statement in its containing block because the "body" of the `let` are all of the statements in which the variable it binds is in scope. The other cases handle when the `let` comes first, and when the block has no `let` at all.

Reply:
Ahh yes, that makes sense, thanks!